library/runtime_arm64.c

```diff
@@ -0,0 +1,98 @@
+/*
+ *  Runtime detection module for Arm64 CPU feature sets.
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 ( the "License" ); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+#include "runtime_internal.h"
+
+#if defined(MBEDTLS_RUNTIME_C) && \
+    defined(MBEDTLS_ARCH_IS_ARM64) && \
+    !defined(MBEDTLS_CPU_HAS_FEATURES_ALT)
+
+#include <stdint.h>
+
+#if defined(__linux__)
+#include <sys/auxv.h>
+#else
+#include <arm_acle.h>
+#endif
+
+#if defined(__linux__)
+
+bool mbedtls_cpu_has_features_arm64(unsigned long hwcap, unsigned long hwcap2)
+{
+    if ((getauxval(AT_HWCAP) & hwcap) != hwcap) {
+        return false;
+    }
+    return (getauxval(AT_HWCAP2) & hwcap2) == hwcap2;
+}
+
+#else
+/* Access ID_*_EL1 to get the CPU feature sets.
+ *
+ * See: https://github.com/torvalds/linux/blob/master/Documentation/arm64/elf_hwcaps.rst
+ *
+ * Warning: If the register can not be access, it might raise illegal error.
+ */
+static unsigned long cpu_features = 0;
+
+#define MBEDTLS_READ_REG(reg) __arm_rsr64(#reg)
+#define MBEDTLS_GET_BITS(val, start, len) (((val) >> start) & ((1ULL<<len) -1))
+static void cpu_feature_profile(void)
+{
+    uint64_t value, t;
+
+    if (cpu_features > UINT32_MAX) {
+        return;
+    }
+
+    value = MBEDTLS_READ_REG(ID_AA64ISAR0_EL1);
+
+    t = MBEDTLS_GET_BITS(value, 4, 4);
```

Comment:
Magic values like these should be `#define`d

Reply:
> Magic values

I am not sure I understand this comment. Do you mean define the value of register? like 
```
#define AES_V1 1
#define AES_V2 2
```

Reply:
I'd normally expect to see something like (without thinking too hard about the names)

```
#define ID_AA64ISAR0_EL1_AES_SHIFT   4
#define ID_AA64ISAR0_EL1_AES_BITS     4
```